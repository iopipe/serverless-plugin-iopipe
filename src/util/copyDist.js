const fs = require('fs-extra');
const path = require('path');

const distFolder = path.resolve(__dirname, '../');
const newFolder = path.resolve(
  __dirname,
  '../../example/.serverless_plugins/serverless-plugin-iopipe'
);

//only copy the dist folder if needed for local development
if (process.env.LOCAL_PLUGIN) {
  fs.copySync(distFolder, newFolder);
}
