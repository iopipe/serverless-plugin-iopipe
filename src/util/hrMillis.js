import _ from 'lodash';

export default function(startTime = process.hrtime(), suffix = true) {
  const endTime = process.hrtime(startTime);
  const millis = _.round(endTime[0] * 1000 + endTime[1] / 1000000);
  return suffix ? `${millis}ms` : millis;
}
