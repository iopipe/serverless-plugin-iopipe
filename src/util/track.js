import ua from 'universal-analytics';
import fs from 'fs-extra';
import _ from 'lodash';
import { join } from 'path';
import uuid from 'uuid';
import { createHash } from 'crypto';
import { default as debugLib } from 'debug';

import options from '../options';

let userId = undefined;
let visitor = undefined;

const debug = debugLib('serverless-plugin-iopipe:track');

export function set(instance) {
  // create consistent, yet anonymized id for usage stats
  const pkg = fs.readJsonSync(join(instance.prefix, 'package.json'));
  const str =
    pkg.author ||
    _.get(pkg, 'repository.url') ||
    pkg.name ||
    pkg.homepage ||
    uuid.v4();
  userId = createHash('md5').update(str).digest('hex');
  visitor = ua('UA-73165042-2', userId, {
    strictCidFormat: false,
    https: true
  });
}

export function track(obj = {}) {
  if (!visitor) {
    return Promise.resolve('no-visitor');
  }
  if (options().noStats) {
    return Promise.resolve('no-stats');
  }
  const { category = 'event', action = 'action', label = 'label', value } = obj;
  const newLabel = _.isString(label) ? label : JSON.stringify(label);
  debug(`Tracking ${category}: ${action}`);
  return new Promise((resolve, reject) => {
    visitor.event(category, action, newLabel, value, (err, res) => {
      return err ? reject(err) : resolve(res);
    });
  });
}
