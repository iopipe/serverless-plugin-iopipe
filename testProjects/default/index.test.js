/*eslint-disable import/no-extraneous-dependencies*/
import _ from 'lodash';

import { cleanup, run, unzip } from '../util/unzipRun';

process.env.IOPIPE_TOKEN = 'test_token';

const dir = __dirname;

beforeAll(() => {
  unzip({ dir });
});

afterAll(() => {
  cleanup({ dir });
});

test('Generated files require plugin, include plugin inline, and export original handler', async () => {
  const simpleRes = await run({
    dir,
    file: 'simple-iopipe.js'
  });
  expect(simpleRes.statusCode).toBe(200);

  const nameMismatchRes = await run({
    dir,
    file: 'nameMismatch-iopipe.js'
  });
  expect(nameMismatchRes).toBe(301);

  const syntaxErrorRes = await run({
    dir,
    file: 'syntaxError-iopipe.js'
  });
  expect(syntaxErrorRes.message).toMatch(/Unexpected\stoken,\s/);
});
