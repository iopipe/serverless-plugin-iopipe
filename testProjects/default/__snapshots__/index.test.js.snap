// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated files require plugin, include plugin inline, and export original handler 1`] = `
"const iopipe = require('iopipe')({\\"token\\":\\"test-token\\",\\"installMethod\\":\\"serverless-plugin-iopipe@0.0.0-semantically-released\\",\\"plugins\\":[]});

let handler, handlerError;
// The following is an automatically generated require statement by the plugin,
// aimed to provide syntax/type errors to the IOpipe service.
// The original file is imported as text with capitalized tokens replaced.
try {
  handler = require('../handlers/simple');
} catch (err) {
  handlerError = err;
}

exports['simple'] = function attemptSimple(event, context, callback) {
  try {
    return iopipe((evt, ctx, cb) => {
      if (handlerError) {
        return cb(handlerError);
      }
      return handler.handler(evt, ctx, cb);
    })(event, context, callback);
  } catch (err) {
    throw err;
  }
};
"
`;

exports[`Generated files require plugin, include plugin inline, and export original handler 2`] = `
"const iopipe = require('iopipe')({\\"token\\":\\"test-token\\",\\"installMethod\\":\\"serverless-plugin-iopipe@0.0.0-semantically-released\\",\\"plugins\\":[]});

let handler, handlerError;
// The following is an automatically generated require statement by the plugin,
// aimed to provide syntax/type errors to the IOpipe service.
// The original file is imported as text with capitalized tokens replaced.
try {
  handler = require('../handlers/differentName');
} catch (err) {
  handlerError = err;
}

exports['nameMismatch'] = function attemptNameMismatch(event, context, callback) {
  try {
    return iopipe((evt, ctx, cb) => {
      if (handlerError) {
        return cb(handlerError);
      }
      return handler.wow(evt, ctx, cb);
    })(event, context, callback);
  } catch (err) {
    throw err;
  }
};
"
`;

exports[`Generated files require plugin, include plugin inline, and export original handler 3`] = `
"const iopipe = require('iopipe')({\\"token\\":\\"test-token\\",\\"installMethod\\":\\"serverless-plugin-iopipe@0.0.0-semantically-released\\",\\"plugins\\":[]});

let handler, handlerError;
// The following is an automatically generated require statement by the plugin,
// aimed to provide syntax/type errors to the IOpipe service.
// The original file is imported as text with capitalized tokens replaced.
try {
  handler = require('../handlers/syntaxError');
} catch (err) {
  handlerError = err;
}

exports['syntaxError'] = function attemptSyntaxError(event, context, callback) {
  try {
    return iopipe((evt, ctx, cb) => {
      if (handlerError) {
        return cb(handlerError);
      }
      return handler.handler(evt, ctx, cb);
    })(event, context, callback);
  } catch (err) {
    throw err;
  }
};
"
`;
